cpp/simulations/abm.cpp

```diff
@@ -849,35 +841,95 @@ int main()
-    // plot "abm.txt" using 1:5 with lines title "infected (detected)", "abm.txt" using 1:10 with lines title "dead"
-    // set xlabel "days"
-    // set ylabel "number of people"
-    // set title "ABM Example"
-    // set output "abm.png"
-    // set terminal png
-    // replot
-    auto f_abm = fopen("abm.txt", "w");
-    fprintf(f_abm, "# t S E C I I_s I_c R_C R_I D\n");
-    for (auto i = 0; i < sim.get_result().get_num_time_points(); ++i) {
-        fprintf(f_abm, "%f ", sim.get_result().get_time(i));
-        auto v = sim.get_result().get_value(i);
-        for (auto j = 0; j < v.size(); ++j) {
-            fprintf(f_abm, "%f", v[j]);
-            if (j < v.size() - 1) {
-                fprintf(f_abm, " ");
+/**
+ * Run the abm simulation
+ * @param result_dir directory where all results of the parameter study will be stored.
+ * @param num_runs number of runs.
+ * @param save_single_runs [Default: true] Defines if single run results are written to the disk.
+ * @returns any io error that occurs during reading or writing of files.
+ */
+mio::IOResult<void> run(const fs::path& result_dir, size_t num_runs, bool save_single_runs = true)
+{
+
+    auto t0         = mio::abm::TimePoint(0); // Start time (t0) per simulation
+    auto tmax       = mio::abm::TimePoint(0) + mio::abm::days(60); // Endtime (tmax) per simulation
+    auto ensemble_results = std::vector<std::vector<mio::TimeSeries<double>>>{}; // Vector of collected results (TimeSeries objects)
+    ensemble_results.reserve(size_t(num_runs));
+    auto run_idx            = size_t(1); //The run index
+    auto save_result_result = mio::IOResult<void>(mio::success());
+
+    // Loop over a number of run
+    while (run_idx <= num_runs) {
+        
+        // Create the sampled simulation with start time t0.
+        auto sim = create_sampled_simulation(t0);
+        // Collect the id of location in world.
+        std::vector<int> loc_ids;
+        for (auto&& locations : sim.get_world().get_locations()) {
+            for (auto location : locations){
+                loc_ids.push_back(location.get_index());
             }
         }
-        if (i < sim.get_result().get_num_time_points() - 1) {
-            fprintf(f_abm, "\n");
+        // Advance the world to tmax
+        sim.advance(tmax);
+        // TODO: update result of the simulation to be a vector of location result.
+        auto temp_sim_result = std::vector<mio::TimeSeries<double>>{sim.get_result()};
+        // Push result of the simulation back to the result vector
+        ensemble_results.push_back(temp_sim_result);
+        // Option to save the current run result to file
+        if (save_result_result && save_single_runs) {
+            auto result_dir_run = result_dir / ("abm_result_run_" + std::to_string(run_idx) + ".h5");
+            BOOST_OUTCOME_TRY(save_result(ensemble_results.back(), loc_ids, (int)mio::abm::InfectionState::Count, result_dir_run.string()));
```

Comment:
```suggestion
            BOOST_OUTCOME_TRY(save_result(ensemble_results.back(), loc_ids, 1, result_dir_run.string()));
```
Here, we now use the compartments/infection states as Groups. In the other models contexts, a group would be, e.g., an AgeGroup and inside this group we have all the infection states of this particular age. To be consistent we should here have int num_groups=1 (third parameter) temporarily and store our results age resolved at some later point with using the third parameter > 1.

Reply:
Yes. I didn't know the meaning of the group for this function so made assumption that it represents the subpopulations. 
It is now resolved accordingly.